File: CIMToolPlugin/src/au/com/langdale/ui/binding/TableBinding.java
/*
 * This software is Copyright 2005,2006,2007,2008 Langdale Consultants.
 * Langdale Consultants can be contacted at: http://www.langdale.com.au
 */
package au.com.langdale.ui.binding;

import org.eclipse.jface.viewers.CheckboxTableViewer;
import org.eclipse.jface.viewers.StructuredViewer;

import au.com.langdale.ui.plumbing.Binding;
import au.com.langdale.ui.plumbing.Plumbing;
/**
 * A binding that displays an array as the checked items in a table viewer.
 */
public abstract class TableBinding implements Binding, ArrayModel, AnyModel {
	private Object[] values = new Object[0];
	private CheckboxTableViewer viewer;
	private Plumbing plumbing;
	private AnyModel parent;
	
	public void bind(String name, Plumbing plumbing, AnyModel parent) {
		this.plumbing = plumbing;
		this.parent = parent;
		viewer = (CheckboxTableViewer) plumbing.getViewer(name);
		configureViewer(viewer);
		plumbing.addBinding(this, parent);
	}
	
	public void bind(String name, Plumbing plumbing) {
		bind(name, plumbing, null);
	}

	protected abstract void configureViewer(StructuredViewer viewer);
	protected abstract Object getInput();

	public AnyModel getParent() {
		return parent;
	}

	protected void resetInput(StructuredViewer viewer) {
		Object input = getInput();
		if( input == null || ! input.equals(viewer.getInput()))
			viewer.setInput(input);
	}

	public Object[] getValues() {
		return values;
	}

	public void setValues(Object[] values) {
		this.values = values;
		if( plumbing != null)
			plumbing.doRefresh();
	}
	
	public Object getValue() {
		if( values.length > 0)
			return values[0];
		else
			return null;
	}

	public void setValue(Object value) {
		setValues((value != null? new Object[] { value } : new Object[0]));
	}

	public void refresh() {
		resetInput(viewer);
		if(viewer.getInput() != null)
			viewer.setCheckedElements(values);
	}

	public void reset() {
		viewer.setAllChecked(false);
	}

	public void update() {
		resetInput(viewer);
		values = viewer.getCheckedElements();
	}

	public String validate() {
		return null;
	}
}

File: ScalaRCP/src/au/com/langdale/eclipse/TableBinding.scala
package au.com.langdale
package eclipse
import ui.binding.{AnyModel, ArrayModel, ListBinding, TableBinding => ChecksBinding}
import ui.plumbing.{Binding}
import ui.builder.{Assembly}
import ui.util.IconCache
import util.Descriptor

import org.eclipse.jface.viewers.{StructuredViewer,Viewer,TableViewer,TableViewerColumn,LabelProvider,ITableLabelProvider,IStructuredContentProvider}
import org.eclipse.swt.SWT

object TableBinding {
  def apply[T <: AnyRef](cols: Descriptor[T], checkboxes: Boolean) =
    if( checkboxes )
      new ChecksBinding with TableBinding[T] { val descr = cols }
    else
      new ListBinding with TableBinding[T] { 
        val descr = cols
        var plumbing: Assembly = _
        
        def getValues: Array[AnyRef] = 
          if( getValue == null ) 
            Array()
          else
            Array(getValue)
        
        def setValues(values: Array[AnyRef]) {
          if( values.isEmpty )
            setValue(null)
          else
            setValue(values(0))
          plumbing.doRefresh
        }
        
        override def bind( name: String, assembly: Assembly) {
          plumbing = assembly
          super.bind(name, assembly)
        }
      }
  
  
}


trait TableBinding[T <: AnyRef] extends Binding with AnyModel with ArrayModel  {
  val descr: Descriptor[T]
  
  private var input: Seq[AnyRef] = Nil
  private lazy val props = descr.toList.toArray
  
  def bind( name: String, assembly: Assembly)
  def getInput = input
  def setInput( given: Seq[AnyRef]) { input = given }

  def configureViewer(raw: StructuredViewer) {
     raw match {
       case viewer: TableViewer =>
         for( prop <- props ) {
           val vcol = new TableViewerColumn(viewer, SWT.NONE)
           val tcol = vcol.getColumn
           tcol setText prop.desc
           tcol setResizable true
           tcol setMoveable true
           tcol setWidth 100
         }
         val table = viewer.getTable
         table setHeaderVisible true
         table setLinesVisible true
         viewer setLabelProvider labels
         viewer setContentProvider provider
     }
  }
  
  private def truncate(text: String) = {
    val ix = text.indexOf("\n")
    val iy = if( ix == -1 ) if(text.length > 50) 50 else text.length else ix
    text.substring(0, iy)
  }
  
  private object labels extends LabelProvider with ITableLabelProvider {
    def getColumnText( raw: Any, ix: Int) = props(ix).get(raw.asInstanceOf[T]) match {
      case null => ""
      case Some(value) => truncate(value.toString.trim)
      case None => ""
      case value => truncate(value.toString.trim)
    } 
    def getColumnImage( raw: Any, ix: Int) = if( ix == 0 ) IconCache.getIcons.get(raw) else null
  }
  
  private object provider extends IStructuredContentProvider {
    def getElements(i: Any) = input.toArray[AnyRef]
    def inputChanged(v: Viewer, o: Any, n: Any) {}
    def dispose {}
  }
  
  def selection = getValues.map( _.asInstanceOf[T] ).toList
  def contents = getInput.map( _.asInstanceOf[T] ).toList
}
